mpeg2/src/temi.rs

```diff
@@ -0,0 +1,271 @@
+use crate::bitstream::{Bitstream, BitstreamWriter, Decode};
+use crate::{DecodeError, EncodeError};
+
+pub const AF_DESCR_TAG_TIMELINE: u8 = 0x04;
+const TEMI_TIMELINE_FLAGS_LENGTH: usize = 3;
+
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
+pub enum TimeFieldLength {
+    None = 0,
+    Short = 1,
+    Long = 2,
+    Reserved = 3,
+}
+
+impl Default for TimeFieldLength {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct TEMITimelineDescriptor {
+    pub has_timestamp: TimeFieldLength,
+    pub timescale: u32,
+    pub media_timestamp: u64,
+
+    pub ntp_timestamp: Option<u64>,
+    pub ptp_timestamp: Option<u128>,
+
+    pub drop: bool,
+    pub frames_per_tc_seconds: u16,
+    pub duration: u16,
+    pub has_timecode: TimeFieldLength,
+    pub time_code: u64,
+
+    pub force_reload: bool,
+    pub paused: bool,
+    pub discontinuity: bool,
+    pub timeline_id: u8,
+}
+
+impl TEMITimelineDescriptor {
+    pub fn encoded_len(&self) -> usize {
+        let mut len = if self.has_timestamp == TimeFieldLength::Short {
+            13
+        } else if self.has_timestamp == TimeFieldLength::Long {
+            17
+        } else {
+            5
+        };
+        if self.ntp_timestamp.is_some() {
+            len += 8;
+        }
+        if self.ptp_timestamp.is_some() {
+            len += 10;
+        }
+
+        len += if self.has_timecode == TimeFieldLength::Short {
+            7
+        } else if self.has_timecode == TimeFieldLength::Long {
+            12
+        } else {
+            0
+        };
+        len
+    }
+
+    pub fn encode(&self, bs: &mut BitstreamWriter) -> Result<usize, EncodeError> {
```

Comment:
It might be kind of nice to make `Bitstream` an implementation detail instead of a public API:

```suggestion
    pub fn encode(&self, buf: &[u8]) -> Result<usize, EncodeError> {
        let bs = Bitstream::new(buf);
```